server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -812,7 +812,8 @@ String computeResultLevelCachingEtag(
       Hasher hasher = Hashing.sha1().newHasher();
       boolean hasOnlyHistoricalSegments = true;
       for (SegmentServerSelector p : segments) {
-        if (!p.getServer().pick().getServer().isSegmentReplicationTarget()) {
+        QueryableDruidServer queryableServer = p.getServer().pick(query);
+        if (queryableServer == null || !queryableServer.getServer().isSegmentReplicationTarget()) {
```

Comment:
nit: I think it should ideally fail if `queryableServer` is null, because it means there is no server to process this segment. However, this seems OK because this method will return immediately when it's null and fail at [here](https://github.com/apache/druid/pull/10428/files#diff-d0873a0b1d76e92318ed707d876a168cf8ad974a4c53ebe65d7a39ab3cf0294eR594).